### DISCUSSION TITLE
"Target class [livewire] does not exist" issue with Laravel package on Laravel 9

### DISCUSSION THREAD

**Discussion by Okipa** [larastan/larastan#1345] [Q&A] [answered]
Hi there,

I'm developing the v5 of my package [laravel-table](https://github.com/Okipa/laravel-table/tree/v5), which is now based on Livewire.

I'm struggling with the following error when I try to run Larastan : `Target class [livewire] does not exist`.

Everything seems to be correctly configured so I have no clue about what is causing this error.

Everything is going well when running other analysis (Pint, PHPMD, PHPUnit) or when I use the package in a project.

The error can only be reproduced by cloning the package https://github.com/Okipa/laravel-table/tree/v5, installing composer dependencies with `composer install`, switching on branch v5 with a `git checkout v5` and then running PHPStan with `vendor/bin/phpstan analyse --memory-limit=2G`.

I would like to release the v5 in stable version but I would be way more confident if I could launch PHPStan analysis locally and in CI. I'm searching for quite a time for similar problems and I found nothing useful to fix this issue.

Could you at least point me in the right direction? Thanks by advance for your lights!

EDIT: I should precise that this package is compatible with PHP8.1 and Laravel 8 or 9.

**Comment by Okipa** [comment]
For information, I also opened an issue on the `livewire` repo but it had no success for now: https://github.com/livewire/livewire/discussions/5117

**Comment by szepeviktor** [comment]
Hello Arthur!!
All I can think of
1. `LiveWire` is a facade
2. It cannot be resolved from the app container

**Reply by Okipa** [reply]
Hi @szepeviktor, thanks for your answer.

Are you telling me that Larastan can't be used with packages based on Livewire?

Or would you see any solution allowing it to work in this specific case?

**Reply by szepeviktor** [reply]
Temporarily please insert
`\Illuminate\Support\Facades\App::bind('livewire', function () { return new \Livewire\LivewireManager; });`
in `vendor/nunomaduro/larastan/src/ApplicationResolver.php`
right after `$app = (new self)->createApplication();`

🙃

**Reply by szepeviktor** [reply]
... or rather add a method

```php
    protected function overrideApplicationBindings($app)
    {
        return [
            'livewire' => 'Livewire\LivewireManager',
        ];
    }
```

**Reply by szepeviktor** [reply]
@canvural I know you can come up with a normal solution.

**Reply by Okipa** [reply]
@szepeviktor, thanks ! It does temporarily fix the issue 👍 

@canvural Would you see a way to handle this case properly ? 

I had this suggestion on the Livewire repo but it does not completely fix the issue, PHPStan does run but the error still shows at the end of the analysis: https://github.com/livewire/livewire/discussions/5117#discussioncomment-3494484

**Comment by szepeviktor** [comment]
Some results: https://github.com/Okipa/laravel-table/pull/98#issuecomment-1230309188

**Comment by yajra** [comment]
Similar issue, using the app helper instead of the facade to register a component works for me.

```php
        if (! $this->app->runningInConsole()) {
            app('livewire')->component('export-button', ExportButtonComponent::class);
        }
```

Also, downgrading to 2.1.12 fixes it. The issue seems to be on 2.2.0 release.

**Reply by Okipa** [reply]
It seems that it doesn't change anything.

Since the 2.2 version, the analysis does run and returns this kind of error (for each test) at the end:

```
Internal error: Internal error: Target class [livewire] does not exist. in file /Users/arthurlorent/code/laravel-table/src/LaravelTableServiceProvider.php
```

**Reply by szepeviktor** [reply]
@canvural Could we revert/fix the thing that was introduced in v2.2?

**Reply by Okipa** [reply]
@szepeviktor Well, before the v2 I had this error:

```
Illuminate\Contracts\Container\BindingResolutionException thrown in /Users/arthurlorent/code/laravel-table/vendor/laravel/framework/src/Illuminate/Container/Container.php on line 877 while loading bootstrap file /Users/arthurlorent/code/laravel-table/vendor/nunomaduro/larastan/bootstrap.php: Target class [livewire] does not exist.
```

At least the analysis is running now, but there still is something wrong between Livewire and Larastan that prevent analysis to run normally.

/cc @canvural

**Reply by canvural** [reply]
We use `testbench` to boot up a Laravel application during the analysis. And if I clone your repo and run `./vendor/bin/testbench package:discover` I also get `Class "livewire" does not exist ` Which got me thinking maybe the issue is in Testbench somehow. @crynobone Do you have any idea why `livewire` can't be found during `./vendor/bin/testbench package:discover`?

**Reply by szepeviktor** [reply]
> I also get `Class "livewire" does not exist`

The cause can be read here https://github.com/Okipa/laravel-table/pull/98#issuecomment-1230309188

**Comment by canvural** [comment, marked as answer]
@Okipa Just released https://github.com/nunomaduro/larastan/releases/tag/2.2.3 with the fix. Please test it.

**Reply by Okipa** [reply]
@canvural Sorry for the late answer, it works, thanks !
